Extract archive on macos

bensejas opened the pull request:
Adding support to extract an archive file on macOS agents. `bsdtar` is the default `tar` on macOS and its command options differs from `gnu-tar` which is the default on ubuntu.

arromer reviewed (approved):
LGTM 👌

RayBrokeSomething commented on hooks/checkout:
```diff
@@ -139,18 +165,14 @@ copy_checkout_from_s3() {
     aws s3 cp "${s3_url}/${checkout}" "${PWD}/${checkout}"
```
```suggestion
    aws s3 cp "${s3_url}/${checkout}" "${PWD}/${checkout}" --silent
```

RayBrokeSomething replied:
```suggestion
    aws s3 cp "${s3_url}/${checkout}" "${PWD}/${checkout}" --silent
```

RayBrokeSomething commented on hooks/checkout:
```diff
@@ -136,21 +162,17 @@ copy_checkout_from_s3() {
     aria2c --max-connection-per-server=16 --split=16 "${signed_url}" --out="${checkout}"
   else
-    aws s3 cp "${s3_url}/${checkout}" "${PWD}/${checkout}"
+    aws s3 cp "${s3_url}/${checkout}" "${PWD}/${checkout}" --silent
```
```suggestion
    aws s3 cp "${s3_url}/${checkout}" "${PWD}/${checkout}" --quiet
```

oops messed it up

RayBrokeSomething replied:
Actually can fully remove this, it didn't do what I thought it would do... stop logging the download percentage.

bensejas commented:
@arromer I'm going to need another review. I added 2 changes after the initial approval:
1. Removed `pushd .. > /dev/null` and `popd` to stay within the working directory when extracting the S3 archive.
2. Attempt to change the directory to the first child of the extracted archive when the extracted archive is not a git repository. The next steps expect the current directory to be a git respository.

arromer commented on hooks/checkout:
```diff
@@ -139,18 +165,14 @@ copy_checkout_from_s3() {
-  popd >/dev/null
+
+  # Check if extracted archive is a repository
+  if [[ "$(git rev-parse --is-inside-work-tree 2>/dev/null)" != "true" || -n "$(git check-ignore .)" ]]; then
+    # Change directory to first child when repository is not found
+    pushd $(ls -d */ | head -n 1)
+  fi
```
when would this be required?
I don't think this is going to work since the repo would then be located in a different directory to where buildkite will try to use.
